Source/WebKit/UIProcess/geoclue/GeoclueGeolocationProvider.cpp

```diff
@@ -90,45 +86,268 @@ void GeoclueGeolocationProvider::stop()
+            GRefPtr<GVariant> locationData;
+
+            g_variant_get(parameters, "(o@a{sv})", nullptr, &locationData.outPtr());
+
+            auto& provider = *static_cast<GeoclueGeolocationProvider*>(userData);
+            provider.portalLocationUpdated(locationData.get());
+        }, this, nullptr);
+
+    GVariantBuilder options;
+    g_variant_builder_init(&options, G_VARIANT_TYPE_VARDICT);
+    g_variant_builder_add(&options, "{sv}", "handle_token", g_variant_new_string(token.ascii().data()));
+
+    g_dbus_proxy_call(m_portal.locationPortal.get(), "Start", g_variant_new("(osa{sv})", m_portal.sessionId->ascii().data(), "", &options), G_DBUS_CALL_FLAGS_NONE, -1, m_cancellable.get(),
+        [](GObject* manager, GAsyncResult* result, gpointer userData) {
+            GUniqueOutPtr<GError> error;
+            GRefPtr<GVariant> returnValue = adoptGRef(g_dbus_proxy_call_finish(G_DBUS_PROXY(manager), result, &error.outPtr()));
+            if (g_error_matches(error.get(), G_IO_ERROR, G_IO_ERROR_CANCELLED))
+                return;
+
+            auto& provider = *static_cast<GeoclueGeolocationProvider*>(userData);
+            if (error)
+                provider.didFail(_("Failed to connect to geolocation service"));
+        }, this);
+}
+
+void GeoclueGeolocationProvider::portalLocationUpdated(GVariant* variant)
+{
+    WebCore::GeolocationPositionData position;
+
+    g_variant_lookup(variant, "Accuracy", "d", &position.accuracy);
+    g_variant_lookup(variant, "Altitude", "d", &position.altitude);
+    g_variant_lookup(variant, "Heading", "d", &position.heading);
+    g_variant_lookup(variant, "Latitude", "d", &position.latitude);
+    g_variant_lookup(variant, "Longitude", "d", &position.longitude);
+    g_variant_lookup(variant, "Speed", "d", &position.speed);
+
+    guint64 timestamp;
+    if (g_variant_lookup(variant, "Timestamp", "(tt)", &timestamp, nullptr))
+        position.timestamp = static_cast<double>(timestamp);
+
+    m_updateNotifyFunction(WTFMove(position), std::nullopt);
+}
+
+void GeoclueGeolocationProvider::stopPortalSession()
+{
+    if (!m_portal.sessionId)
+        return;
+
+    ASSERT(m_portal.locationPortal);
+
+    auto* connection = g_dbus_proxy_get_connection(m_portal.locationPortal.get());
+    ASSERT(connection);
+
+    if (m_portal.locationUpdatedSignalId) {
+        g_dbus_connection_signal_unsubscribe(connection, m_portal.locationUpdatedSignalId);
+        m_portal.locationUpdatedSignalId = 0;
+    }
+
+    if (m_portal.responseSignalId) {
+        g_dbus_connection_signal_unsubscribe(connection, m_portal.responseSignalId);
+        m_portal.responseSignalId = 0;
+    }
+
+    m_portal.senderName = std::nullopt;
+
+    g_dbus_connection_call(connection, "org.freedesktop.portal.Desktop", m_portal.sessionId->ascii().data(),
+        "org.freedesktop.portal.Session", "Close", nullptr, nullptr, G_DBUS_CALL_FLAGS_NONE, -1, nullptr, nullptr, nullptr);
```

Comment:
You don't think it's worth adding a small callback just to print any potential error? Could be useful in the future if something goes unexpectedly wrong. (I wouldn't pass the GeoclueGeolocationProvider since you surely don't want to keep it alive for the callback or bother with the cancellable here. I would just print a warning if the call fails and do nothing else.)